<a href="https://colab.research.google.com/github/KacperKaszuba0608/Projects-python/blob/main/Guided_Project_Building_Fast_Queries_on_a_CSV.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Guided Project:<br>Building Fast Queries on a CSV

In this project I'm going to build fast queries on a csv file. I use file from Kaggle about prices of laptops. This data are available at this [link](https://www.kaggle.com/datasets/mohidabdulrehman/laptop-price-dataset). Below columns description:

* ID: A unique identifier for the laptop.
* Company: The name of the company that produces the laptop.
* Product: The name of the laptop.
* TypeName: The type of laptop.
* Inches: The size of the screen in inches.
* ScreenResolution: The resolution of the screen.
* CPU: The laptop CPU.
* RAM: The amount of RAM in the laptop.
* Memory: The size of the hard drive.
* GPU: The graphics card name.
* OpSys: The name of the operating system.
* Weight: The laptop weight.
* Price: The price of the laptop.

## Importing necessary libraries and reading in data

In [ ]:
!wget https://raw.githubusercontent.com/KacperKaszuba0608/Datasets/main/laptops.csv -qO laptops.csv

In [ ]:
import csv

with open('laptops.csv') as f:
    data = csv.reader(f)
    data = list(data)
    header = data[0]
    rows = data[1:]
    
print(header)
print(rows[:5])

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']
[['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'], ['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'], ['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'], ['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'], ['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB 

As we see our data has correct names of columns and values in rows are very different.

## Goal of the project

The goal of this guided project is to create a class that represents our inventory. The methods in that class will implement the queries that we want to answer about our inventory. We will also preprocess that data to make those queries run faster. 

Three queries that we will want to answer:

1. Given a laptop id, find the corresponding data.
2. Given an amount of money, find whether there are two laptops whose total price is that given amount.
3. Identify all laptops whose price falls within a given budget.

## Inventory Class

In [ ]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f: # opening file with data
            data = csv.reader(f) # reading in the file
            data = list(data)
        self.header = data[0] # extracting row with headers
        self.rows = data[1:] # extracting row with values
        for row in self.rows:
            row[-1] = int(row[-1]) # changing type of prices to integer

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.header)
print(len(inventory.rows))

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']
1303


## Finding a Laptop From the Id

In [ ]:
class Inventory():
    import csv
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            data = csv.reader(f)
            data = list(data)
        self.header = data[0]
        self.rows = data[1:]
        for row in self.rows:
            row[-1] = int(row[-1])  
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows: # iterating over all rows 
            if row[0] == laptop_id: # finding correct laptop id
                return row
            else:
                None

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.get_laptop_from_id('3362737'))
print(inventory.get_laptop_from_id('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]
None


Everything is good but we can improve our function to search laptops' ID. This algorithm has time complexity O(R) where R is the number of rows. If we create a dictionary with ID as a keys, we can easily do it much faster.

In [ ]:
class Inventory():
    import csv
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            data = csv.reader(f)
            data = list(data)
        self.header = data[0]
        self.rows = data[1:]
        self.id_to_row = {} #empty dict
        for row in self.rows:
            row[-1] = int(row[-1])
            # creating elements of dict with ID as key
            self.id_to_row[row[0]] = row
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
            else:
                None
                
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row: # checking if the ID is in the dict 
            return self.id_to_row[laptop_id]
        else:
            return None

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.get_laptop_from_id_fast('3362737'))
print(inventory.get_laptop_from_id_fast('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]
None


Now, we are going to compare the performance of this two function:
1. get_laptop_from_id()
2. get_laptop_from_id_fast()

We check time spent performing the function.

In [ ]:
import time
import random

ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]
inventory = Inventory('laptops.csv')

# Times for function with for loop
total_time_no_dict = 0
for id1 in ids:
    start = time.time()
    inventory.get_laptop_from_id(id1)
    end = time.time()
    total_time_no_dict += end-start
    
# Times for function with dictionary
total_time_dict = 0
for id1 in ids:
    start = time.time()
    inventory.get_laptop_from_id_fast(id1)
    end = time.time()
    total_time_dict += end-start
    
print(total_time_no_dict)
print(total_time_dict)

0.7420978546142578
0.0036907196044921875


As we above the result are very different. The result was:

* 0.7420978546142578
* 0.0036907196044921875

If we devide this two results we get that function with dictionary is approximetly 200 time faster.

## Two Laptop Promotion

Our idea is a promotion where you give someone a giht card. This person can use it to buy up to two laptops. Thus in this part of projeckt we are going to write a function which will return a laptop which cost $N$ dollars or two laptops which price is approximetly queals $N$ dollars.

In [ ]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            data = csv.reader(f)
            data = list(data)
        self.header = data[0]
        self.rows = data[1:]
        self.id_to_row = {}
        for row in self.rows:
            row[-1] = int(row[-1])
            self.id_to_row[row[0]] = row
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
            else:
                None
                
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row: 
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotions_dollars(self, dollars):
        # iterate over all rows and check if price == dollars
        for row in self.rows:
            if row[-1] == dollars:
                return True
        # if not iterate over rows 2 times and check if sum od 2 laptops == dollars
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.check_promotions_dollars(1000))
print(inventory.check_promotions_dollars(442))

True
False


As we see, we can buy two or one laptop for 1000\\$ but in our dataset we can't buy laptops for 442\\$. Our solution is good, but we can do it faster. Look, the function which we had written has $O(N^2)$ complexity and we have to iterate over all rows. We can do it faster when we use set to store the unique prices.

In [ ]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            data = csv.reader(f)
            data = list(data)
        self.header = data[0]
        self.rows = data[1:]
        self.id_to_row = {}
        self.prices = set() #set to store prices
        for row in self.rows:
            row[-1] = int(row[-1])
            self.id_to_row[row[0]] = row
            self.prices.add(row[-1]) #adding price to set
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
            else:
                None
                
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row: 
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotions_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False
    
    def check_promotions_dollars_fast(self, dollars):
        if dollars in self.prices: #checking if dollars is in set
            return True
        for price in self.prices: # iterate over prices
            # if devide dollar and price in set return True
            if dollars - price in self.prices:
                return True
        return False

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.check_promotions_dollars_fast(1000))
print(inventory.check_promotions_dollars_fast(442))

True
False


We couldn't see the difference between our two function with our eye but when we check times for them when we upload there a list of prices we will see this difference.

In [ ]:
import random
import time

prices = [random.randint(100, 5000) for _ in range(100)]

time_no_set = 0
for price in prices:
    start = time.time()
    inventory.check_promotions_dollars(price)
    stop = time.time()
    time_no_set += stop - start
    
print(time_no_set)

time_with_set = 0
for price in prices:
    start = time.time()
    inventory.check_promotions_dollars_fast(price)
    stop = time.time()
    time_with_set += stop - start
    
print(time_with_set)

0.7379612922668457
0.0004177093505859375


The results are obvious, second function is much faster as we expected. When we devide this two times, we can say that function with set is approximetly 1844.75 times faster.

## Finding Laptops Within a Budget

Our last goal is to help a customer find all laptops that are in their budget. It means, that customer give us a budget of $N$ dollars and we find all laptops which can be buy by customer. Our function return index of first laptop which is higher then budget. Otherwise function return $-1$ if the is no laptop which price is higher than budget.

In [ ]:
def price_row(row):
    return row[-1]

class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename) as f:
            data = csv.reader(f)
            data = list(data)
        self.header = data[0]
        self.rows = data[1:]
        self.id_to_row = {}
        self.prices = set()
        for row in self.rows:
            row[-1] = int(row[-1])
            self.id_to_row[row[0]] = row
            self.prices.add(row[-1])
        # sorting rows by price
        self.rows_by_price = sorted(self.rows, key = price_row)
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
            else:
                None
                
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row: 
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotions_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False
    
    def check_promotions_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if dollars - price in self.prices:
                return True
        return False
    
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0                                   
        range_end = len(self.rows_by_price) - 1                   
        while range_start < range_end:
            range_middle = (range_end + range_start) // 2  
            price = self.rows_by_price[range_middle][-1]
            if price > target_price:
                range_end = range_middle
            else:
                range_start = range_middle + 1
        if self.rows_by_price[range_start][-1] <= target_price:                  
            return -1                                   
        return range_start

In [ ]:
inventory = Inventory('laptops.csv')
print(inventory.find_first_laptop_more_expensive(1000))
print(inventory.find_first_laptop_more_expensive(10000))

683
-1


At the end of the projeckt I want to add, that preprocesing function is very important in working with data. It can improve  